gdnative-derive/src/export.rs

```diff
@@ -0,0 +1,65 @@
+use proc_macro2::{Span, TokenStream as TokenStream2};
+use syn::spanned::Spanned;
+use syn::{DeriveInput, Fields};
+
+fn err_only_supports_fieldless_enums(span: Span) -> syn::Error {
+    syn::Error::new(span, "#[derive(Export)] only supports fieldless enums")
+}
+
+pub(crate) fn derive_export(input: &DeriveInput) -> syn::Result<TokenStream2> {
+    let derived_enum = match &input.data {
+        syn::Data::Enum(data) => data,
+        syn::Data::Struct(data) => {
+            return Err(err_only_supports_fieldless_enums(data.struct_token.span()));
+        }
+        syn::Data::Union(data) => {
+            return Err(err_only_supports_fieldless_enums(data.union_token.span()));
+        }
+    };
+
+    let export_impl = impl_export(&input.ident, derived_enum)?;
+    Ok(export_impl)
+}
+
+fn impl_export(enum_ty: &syn::Ident, data: &syn::DataEnum) -> syn::Result<TokenStream2> {
+    let err = data
+        .variants
+        .iter()
+        .filter(|variant| !matches!(variant.fields, Fields::Unit))
+        .map(|variant| err_only_supports_fieldless_enums(variant.ident.span()))
+        .reduce(|mut acc, err| {
+            acc.combine(err);
+            acc
+        });
+    if let Some(err) = err {
+        return Err(err);
+    }
+
+    let mappings = data
+        .variants
+        .iter()
+        .map(|variant| {
+            let key = &variant.ident;
+            let val = quote! { #enum_ty::#key as i64 };
+            quote! { (stringify!(#key).to_string(), #val) }
+        })
+        .collect::<Vec<_>>();
+
+    let impl_block = quote! {
+        impl ::gdnative::export::Export for #enum_ty {
+            type Hint = ::gdnative::export::hint::IntHint<i64>;
+            #[inline]
+            fn export_info(hint: Option<Self::Hint>) -> ::gdnative::export::ExportInfo {
+                if let Some(hint) = hint {
+                    return hint.export_info();
```

Comment:
What's the purpose of this bypass path here? I would assume there's not much point in deriving the enum hint but not using it on the user's part. Do you have any examples?

If a type isn't intended to be hinted further, a "never" type should be used for the `type Hint`. More information on this here: https://godot-rust.github.io/docs/gdnative/master/gdnative/export/trait.Export.html#associatedtype.Hint

Reply:
Since `NoHint` is not public, we can't set `type Hint` to it. I think we should make it public? Creating an inhabitable type inside a macro does not seem like a good idea.

Reply:
No. `NoHint` is `#[doc(hidden)] pub` so it can be used in a macro. Remember to use [`proc_macro_crate`](https://docs.rs/proc-macro-crate/latest/proc_macro_crate/) to get the path to the `gdnative` / `gdnative-core` crates. You can find examples from other existing macros.

Reply:
> I don't have any particular problem with creating an inhabitable type within a macro. What do you think is bad about it?

I am not sure how to ensure that a generated type name is unique. Is using input ident as a prefix enough? or should I allow users to override the generated hint name?

Reply:
You don't need a unique name for the type. Just group all the items into a anonymous const scope:

```rust
const _: () = {
    enum __NoHint {}
    impl Export for SomeType {
        type Hint = __NoHint;
        // ...
    }
};
```